### PR TITLE
Adding pylint script to Travis-CI up front.

### LINKED ISSUE #1079: As a NuPIC developer, my builds fail in Travis-CI when pylint fails
From: https://issues.numenta.org/browse/NPC-237

Need to add pylint checking to Travis build. This should run before anything compiles.

### PR DESCRIPTION
**Pull request by rhyolight**
This runs pylint first thing, and fails fast. Only runs on the changed files in the PR, and takes only about a minute to pass/fail.

Fixes #1079. Will post an example soon...

### CODE REVIEW

**Comment by rhyolight** [comment]
See [example output](https://travis-ci.org/numenta/nupic/jobs/131287515#L341-L387) and failure.

```
============= PYLINT =============
===================>
= running pylint on src/nupic/algorithms/CLAClassifier.py
===================>
************* Module CLAClassifier
W0301:435,0: Unnecessary semicolon
C0103:  1,0: Invalid name "CLAClassifier" (should match ([a-z][a-z0-9_]*[a-z0-9])|(__init__)$)
F0401: 28,0: Unable to import 'numpy'
C0103: 40,0:_pFormatArray: Invalid name "array_" (should match _?[a-z][A-Za-z0-9]*$)
C0103: 40,18:_pFormatArray: Invalid name "array_" (should match _{0,2}[a-z][A-Za-z0-9]*$)
W0212:234,21:BitHistory.read: Access to a protected member _stats of a client class
W0212:236,8:BitHistory.read: Access to a protected member _stats of a client class
W0212:238,6:BitHistory.read: Access to a protected member _stats of a client class
W0631:442,19:CLAClassifier.compute: Using possibly undefined loop variable 'learnPatternNZ'
W0212:597,21:CLAClassifier.read: Access to a protected member _learnIteration of a client class
W0212:599,6:CLAClassifier.read: Access to a protected member _patternNZHistory of a client class
W0212:611,8:CLAClassifier.read: Access to a protected member _activeBitHistory of a client class
W0212:618,8:CLAClassifier.read: Access to a protected member _actualValues of a client class
W0212:620,8:CLAClassifier.read: Access to a protected member _actualValues of a client class
W0612:639,9:CLAClassifier.write: Unused variable 'iteration'
===================>
= running pylint on src/nupic/data/record_stream.py
===================>
************* Module record_stream
F0401: 28,0: Unable to import 'nupic.data.fieldmeta'
R0201:278,2:RecordStreamIface.getAggregationMonthsAndSeconds: Method could be a function
W0622:301,43:RecordStreamIface.getRecordsRange: Redefining built-in 'range'
===================>
= running pylint on src/nupic/research/spatial_pooler.py
===================>
************* Module spatial_pooler
C0301:683,0: Line too long (84/80)
C0301:1456,0: Line too long (93/80)
W0511:923,0: TODO: extend to support different number of dimensions for inputs and
W0511:1259,0: TODO: See https://github.com/numenta/nupic.core/issues/128
F0401: 24,0: Unable to import 'numpy'
F0401: 25,0: Unable to import 'nupic.bindings.math'
W0612:329,72:SpatialPooler.__init__: Unused variable 'i'
E0602:684,11:SpatialPooler.setPotential: Undefined variable 'column'
W0212:1730,4:SpatialPooler.read: Access to a protected member _random of a client class
W0212:1762,4:SpatialPooler.read: Access to a protected member _potentialPools of a client class
W0212:1764,4:SpatialPooler.read: Access to a protected member _permanences of a client class
W0212:1768,4:SpatialPooler.read: Access to a protected member _connectedSynapses of a client class
W0212:1770,6:SpatialPooler.read: Access to a protected member _updatePermanencesForColumn of a client class
W0212:1771,8:SpatialPooler.read: Access to a protected member _permanences of a client class
W0201:1637,6:SpatialPooler._seed: Attribute '_random' defined outside __init__
========== PYLINT FAILED =========
```

**Comment by rhyolight** [comment]
@oxtopus What do you think?

**Comment by rhyolight** [comment]
Don't merge this. There is something wrong with it. Somehow the build is stopping (passing) immediately after the pylint process runs instead of continuing.

**Comment by rhyolight** [comment]
Ok, it's better now.

**Comment by breznak** [comment]
Unless whole codebase passes pylint (better option), you should :

- Limit pylint on edited lines only, or
- Make it soft constraint

Imagine a 1 line change to a messy file.
Also, whitespace changes make PRs harder to review, how about special flag
for lint PRs that gets merged quickly.

On Thursday, May 19, 2016, Matthew Taylor <notifications@github.com> wrote:
> Ok, it's better now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub

-- 
Marek Otahal :o)

**Comment by rhyolight** [comment]
That's kindof the point! Everyone has to clean up the files!

**Comment by oxtopus** [comment]
👍

File: ci/travis/run_pylint.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env sh
```
Comment (oxtopus) [inline ci/travis/run_pylint.sh]: This should be `#!/bin/bash` for consistency with other files.

**Pull request merged**